### PR TITLE
Improvements to TOM RPC

### PR DESCRIPTION
**Pull request by topliceanu**
TODO: 

- [ ] add tests for the two use cases: configs and transmissions
- [ ] add env vars to control the rate limiting of the FCD and RPC clients

### CODE REVIEW

**Review by pedrovsg — DISMISSED** [review]

File: pkg/monitoring/chain_reader.go [hunk 1]
```diff
@@ -19,69 +23,54 @@ type ChainReader interface {
...
-func NewChainReader(client *pkgClient.Client) ChainReader {
+func NewChainReader(terraConfig TerraConfig, coreLog logger.Logger) ChainReader {
 	return &chainReader{
-		client,
-		sync.Mutex{},
+		terraConfig,
+		coreLog,
 		sync.Mutex{},
+		ratelimit.New(1,
+			ratelimit.Per(5*time.Second), // one request every 5 seconds
```
Comment (connorwstein) [inline pkg/monitoring/chain_reader.go]: spoke offline - would be good to make an env var so we can quickly tune if needed

**Review by connorwstein — DISMISSED** [review]

File: pkg/monitoring/fcdclient/client.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (c *client) GetTxList(ctx context.Context, params GetTxListParams) (Response, error) {
+	_ = c.limiter.Take()
+	query := url.Values{}
+	if params.Account.String() != "" {
+		query.Set("account", params.Account.String())
+	}
+	if params.Limit != 0 {
+		query.Set("limit", strconv.Itoa(params.Offset))
```
Comment (connorwstein) [inline pkg/monitoring/fcdclient/client.go]: offset and limit are swapped

File: pkg/monitoring/fcdclient/client.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	fcdURL     string
+	httpClient *http.Client
+	limiter    ratelimit.Limiter
+}
+
+func (c *client) GetTxList(ctx context.Context, params GetTxListParams) (Response, error) {
+	_ = c.limiter.Take()
+	query := url.Values{}
+	if params.Account.String() != "" {
+		query.Set("account", params.Account.String())
```
Comment (connorwstein) [inline pkg/monitoring/fcdclient/client.go:40]: should be Add's not Set's

File: pkg/monitoring/fcdclient/types.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type Response struct {
+	Txs []Tx `json:"txs"`
+}
+
+type Tx struct {
+	ID     uint64 `json:"id"`
+	Height uint64 `json:"height"`
```
Comment (connorwstein) [inline pkg/monitoring/fcdclient/types.go]: need string types

**Review by connorwstein — COMMENTED** [review]
Used a little script to find some bugs but I think with those fixed it'll work 
```
package main

import (
	"context"
	"fmt"

	sdk "github.com/cosmos/cosmos-sdk/types"
	core "github.com/terra-money/core/types"

	"github.com/smartcontractkit/chainlink-terra/pkg/monitoring/fcdclient"
)

func main() {
	sdkConfig := sdk.GetConfig()
	sdkConfig.SetBech32PrefixForAccount(core.Bech32PrefixAccAddr, core.Bech32PrefixAccPub)
	sdkConfig.Seal()
	client := fcdclient.New("https://fcd.terra.dev", 100)
	addr, err := sdk.AccAddressFromBech32("terra1dre5vgujqex83kc4kw3jr6fc6z8erdsqxlsvhk")
	if err != nil {
		panic(err)
	}
	fmt.Println(addr.String())
	txes, err := client.GetTxList(context.Background(), fcdclient.GetTxListParams{
		Account: addr,
		Block:   "",
		Offset:  0,
		Limit:   10,
	})
	fmt.Println(txes.Txs[0], err)
}

```

File: pkg/monitoring/fcdclient/types.go (hunk 4 above)
Comment (topliceanu) [reply pkg/monitoring/fcdclient/types.go]: Done

File: pkg/monitoring/fcdclient/client.go (hunk 3 above)
Comment (topliceanu) [reply pkg/monitoring/fcdclient/client.go:40]: At this point setting multiple params is not needed. If/When we do batching, I'll add it!

File: pkg/monitoring/fcdclient/client.go (hunk 2 above)
Comment (topliceanu) [reply pkg/monitoring/fcdclient/client.go]: Done

File: pkg/monitoring/fcdclient/go_generate.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package fcdclient
+
+//go:generate mockery --name Client --output ./mocks/
```
Comment (connorwstein) [inline pkg/monitoring/fcdclient/go_generate.go:3]: can just have this above the interface def

File: pkg/monitoring/source_envelope.go [hunk 6]
```diff
@@ -405,28 +403,37 @@ func (e *envelopeSource) extractDataFromTxResponse(
...
+	for _, tx := range res.Txs {
+		if !strings.Contains(tx.RawLog, fmt.Sprintf(`"type":"%s"`, eventType)) {
 			continue
 		}
-		isMatchingContractAddress := false
-		for _, attribute := range event.Attributes {
-			if attribute.Key == "contract_address" && attribute.Value == contractAddressBech32 {
-				isMatchingContractAddress = true
-				break
+		for _, event := range tx.Logs[0].Events {
```
Comment (connorwstein) [inline pkg/monitoring/source_envelope.go:416]: defensive length check of tx.Logs before grabbing tx.Logs[0] ?

**Review by connorwstein — APPROVED** [review]

File: pkg/monitoring/fcdclient/go_generate.go (hunk 5 above)
Comment (topliceanu) [reply pkg/monitoring/fcdclient/go_generate.go:3]: agreed! but I'm keeping it consistent with other parts of this package. I'll update in a separate PR.

**Pull request merged**